src/transforms/reduce/transform.rs

```diff
@@ -30,52 +30,46 @@ struct ReduceState {
 
 impl ReduceState {
     fn new() -> Self {
-        let fields = HashMap::new();
-        let metadata = EventMetadata::default();
-
         Self {
             events: 0,
             stale_since: Instant::now(),
             creation: Instant::now(),
-            fields,
-            metadata,
+            fields: HashMap::new(),
+            metadata: EventMetadata::default(),
         }
     }
 
     fn add_event(&mut self, e: LogEvent, strategies: &IndexMap<KeyString, MergeStrategy>) {
-        let (value, metadata) = e.into_parts();
-        self.metadata.merge(metadata);
-
-        let fields = if let Value::Object(fields) = value {
-            fields
-        } else {
-            BTreeMap::new()
-        };
-
-        for (k, v) in fields.into_iter() {
-            let strategy = strategies.get(&k);
-            match self.fields.entry(k) {
-                hash_map::Entry::Vacant(entry) => {
-                    if let Some(strat) = strategy {
-                        match get_value_merger(v, strat) {
-                            Ok(m) => {
-                                entry.insert(m);
-                            }
-                            Err(error) => {
-                                warn!(message = "Failed to merge value.", %error);
+        self.metadata.merge(e.metadata().clone());
+
+        if let Some(fields_iter) = e.all_event_fields_skip_array_elements() {
```

Comment:
I don't quite understand why exclude arrays - we could support them with `e.all_event_fields()`, right? I understand that it might be a rare use case, but still possible. Unless I'm missing some implementation details, seems like it might be event simpler to support any path here instead of adding a function to skip arrays.